src/dawn/SIR/SIR.cpp

```diff
@@ -50,6 +193,259 @@ static bool pointerMapComparison(const sir::GlobalVariableMap& map1,
+  }
+  if(Intervals.size() != rhs.Intervals.size()) {
+    output += "[StencilFunction missmatch] Number of Intervals do not match\n";
+    return std::make_pair(output, false);
+  }
+  if(Asts.size() != rhs.Asts.size()) {
+    output += "[StencilFunction missmatch] Size of ASTs does not match\n";
+    return std::make_pair(output, false);
+  }
+  if(!Args.empty() &&
+     !std::equal(Args.begin(), Args.end(), rhs.Args.begin(),
+                 pointeeComparison<sir::StencilFunctionArg>)) {
+    output += "[StencilFunction missmatch] stencil functions arguments do not match\n";
+    for(unsigned i = 0; i < Args.size(); ++i) {
+      auto tmp = pointeeComparisonWithOutput(Args[i], rhs.Args[i]);
+      if(tmp.second == false) {
+        output += "Missmatch of argument " + Args[i]->Name + " " + tmp.first;
+      }
+    }
+    return std::make_pair(output, false);
+  }
+
+  if(!Intervals.empty() &&
+     !std::equal(Intervals.begin(), Intervals.end(), rhs.Intervals.begin(),
+                 pointeeComparison<sir::Interval>)) {
+    output += "[StencilFunction missmatch] stencil functions intervals do not match\n";
+    for(unsigned i = 0; i < Intervals.size(); ++i) {
+      auto tmp = pointeeComparison(Intervals[i], rhs.Intervals[i]);
+      if(tmp == false) {
+        output += "Missmatch of intervals " + Intervals[i]->toString() + " and " +
+                  rhs.Intervals[i]->toString() + "\n";
+      }
+    }
+    return std::make_pair(output, false);
+  }
+
+  if(!Asts.empty()) {
+    auto astcomparison = [](std::shared_ptr<dawn::AST> comparate1,
+                            std::shared_ptr<dawn::AST> comparate2) {
+      return *(comparate1->getRoot()) == *(comparate2->getRoot());
+    };
+    if(!std::equal(Asts.begin(), Asts.end(), rhs.Asts.begin(), astcomparison)) {
+      output += "[StencilFunction missmatch] ASTs do not match\n";
+      for(int i = 0; i < Asts.size(); ++i) {
+        auto out = compareAst(Asts[i], rhs.Asts[i]);
+        if(!out.second) {
+          output += out.first;
+          return std::make_pair(output, false);
+        }
+      }
+    }
+  }
+  return std::make_pair(output, true);
+}
+
+std::pair<std::string, bool>
+sir::StencilFunctionArg::comparison(const sir::StencilFunctionArg& rhs) const {
+  std::string output;
+  if(Name != rhs.Name) {
+    output += "[StencilFunctionArgument missmatch] Names do not match\n";
+    return std::make_pair(output, false);
+  }
+  if(Kind != rhs.Kind) {
+    output += "[StencilFunctionArgument missmatch] Argument Types do not match\n";
+    return std::make_pair(output, false);
+  }
+  return std::make_pair(output, true);
+}
+
+std::pair<std::string, bool> sir::Value::comparison(const sir::Value& rhs) const {
+
+  std::string output;
+  auto type = getType();
+  if(type != rhs.getType()) {
+    output += "[Value missmatch] Values are not of the same type\n";
+    return std::make_pair(output, false);
+  }
+  switch(type) {
+  case sir::Value::TypeKind::Boolean:
```

Comment:
I would write a function which does this kind of stuff (to reduce code duplication).
```
template <class T>
std::pair<std::string, bool> isEqualImpl(const sir::Value& a,  const sir::Value& b, const char* name) {
  if(a.getValue<T> != b.getValue<T>) 
    std::make_pair(dawn::format("[Value mismatch] %s values are not equal\n
                                                   "  left = %s\n"
                                                   "  right = %s\n", name, internal::StdToString<T>()(a), internal::StdToString<T>()(b)), false)
  return std::make_pair(std::string(), true); 
}
```

It might be worth to indenting all the extra information about the members. For example:
```
[Value mismatch] Boolean values are not equal
  left  = true
  right = false
```
seems to be more readable than
```
[Value mismatch] Boolean values are not equal
left  = true
right = false
```
especially if there are multiple reports, you can simply spot the brackets (this may also applies to all the other comparisons).